---
title:  "Antigen Tracking Figures"
author: "Ryan Sheridan"
date:   "`r Sys.Date()`"
output: 
  pdf_document:
    latex_engine: xelatex
    fig_caption: false
documentclass: article
mainfont: Arial
geometry: top=1in,bottom=1in,left=1in,right=1in

params:
  template_dir: "src"
  rds_dir:      "sobjs"
  ref_dir:      "ref"
  # xlsx_dir:     "results/figures"
  # browser_dir:  "results/cellbrowser"
  
  # These parameters specify the input matrices and cell types that should be used for analysis
  # mat_in: Cellranger output folder.
  # cell_type: Cell type used for subsetting Seurat objects.
  # title: Section title.
  # sobj_out: The output Seurat object that is saved after processing and subsetting based on cell type.
  sobjs:
    value:
      d2_DC:
        mat_in:    "results/GEX_CD45pos_d2-ADT_CD45pos_d2/outs/filtered_feature_bc_matrix"
        cell_type: "DC"
        title:     "Myeloid Cells (Day 2)"
        sobj_out:  "walsh_d2_DC.rds"
      d14_DC:
        mat_in:    "results/GEX_CD45pos_d14-ADT_CD45pos_d14/outs/filtered_feature_bc_matrix"
        cell_type: "DC"
        title:     "Myeloid Cells (Day 14)"
        sobj_out:  "walsh_d14_DC.rds"
      d2_LEC:
        mat_in:    "results/GEX_CD45neg_d2-ADT_CD45neg_d2/outs/filtered_feature_bc_matrix"
        cell_type: "LEC"
        title:     "Endothelial/Stromal Cells (Day 2)"
        sobj_out:  "walsh_d2_LEC.rds"
      d14_LEC:
        mat_in:    "results/GEX_CD45neg_d14-ADT_CD45neg_d14/outs/filtered_feature_bc_matrix"
        cell_type: "LEC"
        title:     "Endothelial/Stromal Cells (Day 14)"
        sobj_out:  "walsh_d14_LEC.rds"
      d2_FRC:
        mat_in:    "results/GEX_CD45neg_d2-ADT_CD45neg_d2/outs/filtered_feature_bc_matrix"
        cell_type: "fibroblast"
        title:     "Fibroblast/Stromal Cells (Day 2)"
        sobj_out:  "walsh_d2_FRC.rds"
      d14_FRC:
        mat_in:    "results/GEX_CD45neg_d14-ADT_CD45neg_d14/outs/filtered_feature_bc_matrix"
        cell_type: "fibroblast"
        title:     "Fibroblast/Stromal Cells (Day 14)"
        sobj_out:  "walsh_d14_FRC.rds"
        
  # These parameters specify paths to clustifyr references
  ref_sobjs: ["d2_so.rds", "d14_so.rds"]
  type_ref:  "ref_celltype_walsh.rda"
  xiang_ref: "ref_LEC_xiang.rda"
  subtype_refs:
    value:
      DC:         "ref_DC_walsh.rda"
      LEC:        "ref_LEC_walsh.rda"
      fibroblast: "ref_FRC_walsh.rda"

  abstract: |
    Live, attenuated vaccines generate humoral and cellular immune memory, increasing the duration of protective immune memory. We previously found that antigens derived from vaccination or viral infection persist within lymphatic endothelial cells (LECs) beyond the clearance of the infection, a process we termed “antigen archiving”. Technical limitations of fluorescent labeling have precluded a complete picture of antigen archiving across cell types in the lymph node. We developed a “molecular tracking device” to follow the distribution, acquisition, and retention of antigen in the lymph node. We immunized mice with an antigen conjugated to a nuclease-re-sistant DNA tag and used single-cell mRNA sequencing to quantify its abundance in lymph node hematopoietic and non-hematopoietic cell types. At early and late time points after vaccination, we found antigen acquisition by dendritic cell populations (DCs) and associated ex-pression of genes involved in DC activation and antigen processing, and antigen acquisition and archiving by LECs and unexpected stromal cell types. Variable antigen levels in LECs enabled the identification of caveolar endocytosis as a mechanism of antigen acquisition or retention. Molecular tracking devices enable new approaches to study dynamic tissue dissemination of antigens and identify new mechanisms of antigen acquisition and retention at cellular resolution in vivo.
  methods: |
    FASTQ files from the gene expression and antigen tracking libraries were processed using the cellranger count pipeline (v3.1.0). Reads were aligned to the mm10 and Vaccinia virus (NC_006998) reference genomes. Analysis of gene expression and antigen tracking data was performed using the Seurat R package (v3.2). Antigen tracking and gene expression data were combined into the same Seurat object for each sample (CD45-/day 2, CD45+/day 2, CD45-/day 14, CD45+/day 14). Cells were filtered based on the number of detected genes (>250 and <5000) and the percent of mitochondrial reads (<15%). Gene expression counts were log normalized (NormalizeData), and relative ova signal was calculated by dividing ova-psDNA counts by the median ova-psDNA counts for all T and B cells present in the sample. Gene expression data were scaled and centered (ScaleData). 2000 variable features (FindVariableFeatures) were used for PCA (RunPCA) and the first 40 principal components were used to find clusters (FindNeighbors, FindClusters) and for uniform manifold approximation and projection (UMAP, RunUMAP). Cell types were annotated using the R package [clustifyr](https://rnabioco.github.io/clustifyr) along with reference bulk RNA-seq data from ImmGen (available for download through [clustifyrdata](https://rnabioco.github.io/clustifyrdata)). The samples were then divided into separate objects for DCs, LECs, and FRCs. Cell subsets were then annotated using clustifyr with reference bulk RNA-seq data for DCs, FRCs, and LECs. Ova-low and ova-high populations were identified using a two-component Gaussian mixture model implemented with the R package [mixtools](https://cran.r-project.org/web/packages/mixtools/index.html). Differentially expressed genes were identified using the R package [presto](https://github.com/immunogenomics/presto) and filtered to include those with an adjusted p-value <0.05, log fold change >0.25, AUC >0.5, and with at least 50% of ova-high cells expressing the gene.
---

```{r "Chunk opts", echo = F}

# Default chunk options
knitr::opts_chunk$set(
  message = F,
  warning = F,
  echo    = F
)

```

`r knitr::knit_child(here::here(params$template_dir, "setup.Rmd"))`

```{r "Setup", echo = F}

# Seurat objects used for subtype annotations
# This is needed for Fig S7 heatmaps
sobjs_sub <- read_rds(here(params$rds_dir, "sobjs_subsets.rds"))

# Order of UMAP layers
subtype_order <- list(
  d2_DC   = c("CCR7hi Mig cDC2", "CCR7hi Mig cDC1", "B cell", "T cell"),
  d14_DC  = c("CCR7hi Mig cDC2", "CCR7hi Mig cDC1", "cDC2 Tbet-", "cDC2 Tbet+", "B cell", "T cell", "NK"),
  d2_LEC  = NULL,
  d14_LEC = c("Ptx3_LEC", "fLEC", "cLEC", "Collecting", "BEC", "B cell", "T cell"),
  d2_FRC  = NULL,
  d14_FRC = NULL
)

```

# Abstract

`r params$abstract`

# Methods

`r params$methods`

\newgeometry{top=1in,bottom=1in,left=1.5in,right=1.5in}

# Figure 3

Cell types associated with high antigen counts. (a, d, g, j) UMAP projections are shown for DCs, LECs, and FRCs. (b, e, h, k) Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. (c, f, i, l) Antigen counts are displayed on UMAP projections for each cell type.

```{r "Fig 3", fig.width = 13, fig.height = 18}

# Parameter lists
fig3_names <- c("d2_DC", "d14_DC", "d14_LEC", "d14_FRC")
fig3_sobjs <- sobjs[fig3_names]

fig3_params <- list(
  sobj_in    = fig3_sobjs, 
  plot_title = so_titles[names(fig3_sobjs)],
  cols_in    = so_cols[names(fig3_sobjs)],
  on_top     = subtype_order[names(fig3_sobjs)]
)

# Create panel plots
gg_list <- fig3_params %>%
  pmap(
    create_fig3,
    box_counts   = c("Relative ova signal" = "ova_fc"), 
    umap_counts  = c("ova counts" = "adt_ovalbumin"),
    data_slot    = "counts",
    pt_size_2    = 0.1,
    pt_outline_2 = 0.4
  ) %>%
  flatten()

# Set violin scales equal
violin_idx <- seq_along(gg_list) %% 3 == 2

gg_list[violin_idx] <- gg_list[violin_idx] %>%
  equal_x(log_tran = T)

# Create final figure
fig3 <- gg_list %>%
  wrap_plots(ncol = 3, widths = c(1, 0.75, 1)) +
  plot_annotation(tag_levels = "a") &
  theme(
    plot.margin = unit(c(1, 0.5, 0.5, 0.5), "cm"),
    plot.tag = element_text(size = 24, face = "plain"),
    plot.tag.position = c(-0.08, 1)
  )

fig3

```

\newgeometry{top=1in,bottom=1in,left=2.3in,right=2.3in}

# Figure 4

Genes associated with high antigen counts for day 2 and day 14 cDC2 Tbet- cells. (a, e) Day 2 (a) and day 14 (e) cDC2 Tbet- cells containing low and high antigen counts were identified using a gaussian mixture model. A UMAP projection is shown for ova-low and ova-high cells. Cell types not included in the comparison are shown in white (Other). (b, f) The distribution of ova antigen counts is shown for ova-low and ova-high cDC2 Tbet- cells. Dotted lines indicate the mean counts for each population. (c, g) UMAP projections show the expression of top markers associated with ova-high cDC2 Tbet- cells. (d, h) Expression of ova-high markers for each cell type.

```{r "Fig 4", fig.width = 13, fig.height = 22.1}
# ```{r "Fig 4", fig.width = 10, fig.height = 17}

# Seurat objects, genes, cell types
so_names <- c("d2_DC", "d14_DC")
type <- "cDC2 Tbet-"

feat_cols <- c(
  "#E69F00", "#56B4E9", "#009E73", "#D7301F",
  "#0072B2", "#D55E00", "#6A51A3", "#CC79A7"
)

genes <- list(
  c("Ccl2", "Cxcl2", "Mif", "Msr1", "Pkm", "Lgals3"),
  c("Fcgr4", "Cd72", "Acp5")
) %>%
  set_names(so_names)

genes <- genes %>%
  map(~ set_names(feat_cols[seq_along(.x)], .x))

# OVA colors and groups
fig_cols <- rev(ova_cols)

names(fig_cols) <- names(fig_cols) %>%
  str_c(if_else(. != "Other", str_c(type, "-"), ""), .)
  
gmm_grps <- names(fig_cols)[names(fig_cols) != "Other"]

# Figure parameters
mtplyr_1 <- rep(1, 6)
mtplyr_1[2] <- 2

plot_args <- list(
  sobj_in         = sobjs[so_names],
  feat_cols       = genes[so_names],
  ref_cols        = so_cols[so_names],
  on_top          = subtype_order[so_names],
  plot_labs       = list(letters[1:4], letters[5:8]),
  low_col_mtplyr  = c(2, 1),
  high_col_mtplyr = list(mtplyr_1, 1),
  panel_heights   = list(c(0.6, 1.1, 0.4), c(0.6, 0.6, 0.4)),
  box_pad         = c(0.2, 0.6),
  umap_legd_pad   = c(1.5, 0)
)

# Create figures
fig_4 <- plot_args %>%
  pmap(
    create_fig4,
    ova_cols    = fig_cols,
    gmm_column  = "type_GMM_grp_2",
    gmm_grps    = gmm_grps,
    plot.margin = unit(c(0.2, 0, 0.7, 0), "cm")
  ) %>%
  plot_grid(
    plotlist    = .,    
    rel_heights = c(1.9, 1.4),
    ncol        = 1,
    align       = "vh",
    axis        = "trbl"
  )

fig_4

```

\restoregeometry

# Figure 5

Genes associated with high antigen counts for day 14 LECs. (a) LECs containing low and high antigen counts were identified using a gaussian mixture model. A UMAP projection is shown for ova-low and ova-high cells. T cells, B cells, and epithelial cells are shown in white (Other). (b) Distribution of ova antigen counts for ova-low (light blue) and ova-high (dark blue) cells. Dotted lines indicate the mean counts for each population. (c) The fraction of cells belonging to each LEC cell type for ova-low and ova-high populations. (d) UMAP projections show the expression of top ova-high markers. (e) Expression of ova-high markers for each cell type.

```{r "Fig 5", fig.width = 18, fig.height = 12}
# ```{r "Fig 5", fig.width = 12, fig.height = 8}

# Feature colors
feat_cols <- c(
  "#D7301F", "#0072B2",
  "#009E73", "#4C3250",
  "#E69F00", "#875C04"
)

# Genes of interest
genes <- c("Prox1", "Cavin1", "Cavin2", "Stab1", "Stab2") %>%
  set_names(feat_cols[seq_along(.)], .)

# Color scales for UMAPs
mtplyr <- rep(1, length(genes))
mtplyr[5] <- 2

# Seurat object for figure
so_name <- "d14_LEC"

fig_5 <- create_fig5(
  sobj_in         = sobjs[[so_name]],
  feat_cols       = genes,
  ref_cols        = so_cols[[so_name]],
  ova_cols        = ova_cols,
  pt_size         = 0.2,
  low_col_mtplyr  = 2,
  high_col_mtplyr = mtplyr,
  on_top          = subtype_order[[so_name]],
  gmm_column      = "GMM_grp",
  gmm_grps        = c("ova low", "ova high")
)

fig_5

```

\newgeometry{top=1in,bottom=1in,left=1.7in,right=1.7in}

# Figure S5

Comparison of ova-psDNA, psDNA, and pDNA signal for each cell type. Counts are shown for ova-psDNA, psDNA, and pDNA for DCs (a, b), LECs (c), and FRCs (d).  

```{r "Fig S5", fig.width = 11, fig.height = 17}

box_counts <- c(
  "ovalbumin-psDNA" = "adt_ovalbumin",
  "psDNA"           = "adt_protected-DNA",
  "pDNA"            = "adt_unprotected-DNA"
)

fig_S5 <- sobjs[fig3_names] %>%
  imap(~ {
    create_count_violins(
      sobj_in     = .x,
      plot_cols   = so_cols[[.y]],
      box_counts  = box_counts,
      type_column = "subtype",
      plot_title  = so_titles[[.y]]
    )
  }) %>%
  plot_grid(
    plotlist       = .,
    ncol           = 1,
    labels         = letters,
    label_fontface = "plain",
    label_size     = 18,
    align          = "vh",
    axis           = "trbl"
  )

fig_S5


```

\newgeometry{top=1in,bottom=1in,left=2in,right=2in}

# Figure S6

Antigen counts were compared with total mRNA counts for each cell for DCs (a, b), LECs (c, d), and FRCs (e, f).

```{r "Fig S6", fig.width = 20, fig.height = 34}

# Parameter lists
so_names <- names(sobjs)

corr_params <- list(
  sobj_in    = sobjs[so_names],
  plot_title = so_titles[so_names], 
  cols_in    = so_cols[so_names]
)

# Padding to add to figure panels
pad <- c(9.5, 2, 9.8, 2, 9.8, 2)

# Create scatter plots
cor_plots <- corr_params %>%
  pmap(
    create_figS6,
    x         = c("RNA counts (log10)" = "nCount_RNA"),
    y         = c("ova counts + 1 (log10)" = "adt_ovalbumin"),
    feat      = "subtype",
    split_id  = "subtype",
    data_slot = "counts",
    pt_size   = 1,
    scales    = "free",
    ncol      = 3
  ) %>%
  map2(pad, ~ {
    .x + theme(
      legend.position = "none",
      strip.text      = element_text(size = 14),
      plot.margin     = unit(c(0.2, 0.2, .y, 0.2), "cm")
    )
  })

# Create final figure
figS6 <- plot_grid(
  plotlist       = cor_plots,
  rel_heights    = c(1, 0.8, 0.8),
  labels         = letters,
  label_size     = 32,
  label_fontface = "plain",
  align          = "v",
  ncol           = 2
)

figS6

```

\restoregeometry

# Figure S7

LEC cell types associated wih high antigen counts for the day 2 timepoint. (a) A UMAP projection is shown for LEC cell types, epithelial cells, B cells, and T cells identified for the day 2 timepoint. (b) Relative ova signal is shown for each cell type. Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. (c) Antigen counts are displayed on the UMAP projection shown in (a). (d, e) Correlation coefficients are shown comparing each identified LEC cell type with the reference cell types from Xiang et al. 

```{r "Fig S7", fig.width = 13, fig.height = 9}

# Create UMAPs and violin plots for d2_LEC
so_name <- "d2_LEC"

gg_list <- sobjs[[so_name]] %>%
  create_fig3(
    box_counts   = c("Relative ova signal" = "ova_fc"), 
    umap_counts  = c("ova counts" = "adt_ovalbumin"),
    data_slot    = "counts",
    cols_in      = so_cols[[so_name]],
    plot_title   = so_titles[[so_name]],
    on_top       = subtype_order[[so_name]],
    pt_size      = 0.5,
    pt_outline   = 0.8,
    pt_size_2    = 0.5,
    pt_outline_2 = 0.8
  )

# Set violin scales equal
violin_idx <- seq_along(gg_list) %% 3 == 2
gg_list[violin_idx] <- gg_list[violin_idx] %>%
  equal_x(log_tran = T)

# Create correlation heatmaps for LEC annotations
LEC_names <- c("d2_LEC", "d14_LEC")
heat_cols <- c(get_cols()[2], get_cols()[3])

heat_maps <- map2(LEC_names, heat_cols, ~ {
  create_ref_heatmap(
    sobj_in    = sobjs_sub[[.x]],
    ref_in     = subtype_refs[[so_types[.x]]],
    plot_cols  = .y,
    plot_title = so_titles[[.x]],
    rename_ref = c("BEC (ImmGen)" = "BEC")
  )
})

# Create final figure
fig_S7 <- (gg_list[[1]] | gg_list[[2]] | gg_list[[3]]) /
  (heat_maps[[1]] | heat_maps[[2]]) +
  plot_layout(heights = c(1, 0.6)) +
  plot_annotation(tag_levels = "a") &
  theme(
    plot.margin = unit(c(1, 0.5, 0.5, 0.5), "cm"),
    plot.tag = element_text(size = 24, face = "plain"),
    plot.tag.position = c(0, 1)
  )

fig_S7

```

\newpage

# Figure S9

FRC cell types associated wih high antigen counts for the day 2 timepoint. (a) A UMAP projection is shown for FRC cell types, B cells, and T cells identified for the day 2 timepoint. (b) Relative ova signal is shown for each cell type. Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. (c) Antigen counts are displayed on the UMAP projection shown in (a).

```{r "Fig S9", fig.width = 13, fig.height = 5}

# Create UMAPs and violine plots for d2_FRC
so_name <- "d2_FRC"

gg_list <- sobjs[[so_name]] %>%
  create_fig3(
    box_counts   = c("Relative ova signal" = "ova_fc"), 
    umap_counts  = c("ova counts" = "adt_ovalbumin"),
    data_slot    = "counts",
    cols_in      = so_cols[[so_name]],
    plot_title   = so_titles[[so_name]],
    on_top       = subtype_order[[so_name]],
    pt_size      = 1.5,
    pt_outline   = 2,
    pt_size_2    = 1.5,
    pt_outline_2 = 2
  )

# Set violin scales equal
violin_idx <- seq_along(gg_list) %% 3 == 2
gg_list[violin_idx] <- gg_list[violin_idx] %>%
  equal_x(log_tran = T)

# Create final figure
fig_S9 <- gg_list %>%
  wrap_plots(ncol = 3, widths = c(1, 0.75, 1)) +
  plot_annotation(tag_levels = "a") &
  theme(
    plot.margin = unit(c(1, 0.5, 0.5, 0.5), "cm"),
    plot.tag = element_text(size = 24, face = "plain"),
    plot.tag.position = c(-0.08, 1)
  )

fig_S9

```

\newpage

# Table S2

Comparison of relative ova signal for cell types shown in Fig 3, S7b, and S9b. A two-sided Wilcoxon rank sum test was used to compare relative ova signal for each cell type. Relative ova signal was calculated by dividing antigen counts for each cell by the median antigen counts for T and B cells. The Bonferroni method was used to correct for multiple comparisons. The number of cells in each group (n cells), fraction of the total cells for the sample (frac cells), median relative ova signal, test statistic (statistic), estimation of the median difference (estimate), and confidence interval (conf.low, conf.high) are included.

```{r "Table S2", eval = !is.null(params$xlsx_dir)}

# Create table of ova counts stats
ova_stats <- sobjs %>%
  imap(~ {
    calc_p_vals(
      sobj_in     = .x,
      sample_name = so_titles[.y],
      data_column = "ova_fc",
      type_column = "subtype"
    )
  }) %>%
  bind_rows()

# Write excel file
xlsx_file <- here(params$xlsx_dir, "table_S2.xlsx")

ova_stats %>%
  write_xlsx(
    file_out = xlsx_file,
    sheet_id = "Sample"
  )

```

# Table S3

Genes associated with ova-high cells for DCs, FRCs, and LECs. Ova-high cells were independently identified for DCs, FRCs, and LECs using a gaussian mixture model implemented with the R package mixtools. Differentially expressed genes were identified using a Wilcoxon rank sum test performed using the R package presto (wilcoxauc). The Benjamini-Hochberg method was used to correct for multiple comparisons. Genes were filtered to only include those with an adjusted p-value <0.05, log fold change >0.25, auc >0.5, and with at least 50% of ova-high cells expressing the gene. The average expression, log fold change, test statistic (statistic), area under the receiver operator curve (auc), percentage of ova-high cells that express the gene (pct_in), and percentage of ova-low cells that express the gene (pct_out) are included.

```{r "Table S3"}

# Ova markers
# Need ova_marks for cell browser
ova_marks <- sobjs %>%
  imap(~ {
    find_ova_markers(
      sobj_in    = .x,
      prefix     = "",
      gmm_column = "GMM_grp",
      so_title   = so_titles[.y]
    )
  }) %>%
  bind_rows()

# Write excel file
if (!is.null(params$xlsx_dir)) {
  xlsx_file <- here(params$xlsx_dir, "table_S3.xlsx")
  
  ova_marks %>%
    write_xlsx(
      file_out = xlsx_file,
      sheet_id = "Sample"
    )
}

```

# Table S4

Genes associated with ova-high cells for different DC, FRC, and LEC cell types. Ova-low and ova-high cells were identified independently for each cell type shown in Fig 3 using a gaussian mixture model implemented with the R package mixtools. Differentially expressed genes were identified as described for Supplemental Table 3.

```{r "Table S4", eval = !is.null(params$xlsx_dir)}

# Subtype ova markers
exclude_types <- c("T cell", "B cell", "Epithelial", "NK")

sub_ova_marks <- sobjs %>%
  imap(~ {
    so_in <- .x
    so_title <- so_titles[.y]
    
    cell_types <- so_in@meta.data$subtype %>%
      unique() %>%
      .[!. %in% exclude_types]
    
    res <- cell_types %>%
      map(~ {
        find_ova_markers(
          sobj_in    = so_in,
          prefix     = .x,
          gmm_column = "type_GMM_grp_2",
          so_title   = so_title
        )
      }) %>%
      bind_rows()
    
    res
  }) %>%
  bind_rows()

# Write excel file
xlsx_file <- here(params$xlsx_dir, "table_S4.xlsx")

sub_ova_marks %>%
  write_xlsx(
    file_out = xlsx_file,
    sheet_id = "Sample"
  )

```

# Cell Browser

A UCSC cell browser is available [here](https://d3898ys7yh3545.cloudfront.net)

```{r "Cellbrowser 1", eval = !is.null(params$browser_dir)}

# Save color tsv
browser_cols <- append(so_cols, ova_cols) %>%
  flatten() %>%
  tibble(
    metadataValue = names(.),
    colorCode = .
  ) %>%
  unnest(cols = colorCode) %>%
  unique() %>%
  mutate(colorCode = if_else(
    metadataValue == "Other", 
    "#f0f0f0", 
    colorCode
  ))

# Create cellbrowser directories
sobjs %>%
  iwalk(~ {
    marks_in  <- ova_marks %>%
      filter(Sample == so_titles[[.y]])
    
    create_browser(
      sobj_in       = .x,
      markers       = marks_in,
      browser_dir   = here(params$browser_dir),
      browser_name  = .y,
      cols          = browser_cols,
      browser_title = so_titles[.y],
      abstract      = params$abstract,
      methods_desc  = params$methods
    )
  })

```

```{bash "Cellbrowser 2", eval = FALSE}

html_dir=../cellbrowser

mkdir -p $html_dir/html

browsers=(
  d2_DC  d14_DC
  d2_LEC d14_LEC
  d2_FRC d14_FRC
)

args=""

for dir in ${browsers[@]}
do
  echo -e "\ncolors=\"colors.tsv\"" \
    >> $html_dir/$dir/cellbrowser.conf

  args=$(echo "$args" "-i $html_dir/$dir/cellbrowser.conf")
done

cbBuild $args -o $html_dir/html

aws s3 sync $html_dir/html s3://antigen-tracking-browser

```

